sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+        MethodRequest::GetAccountStateWithProof(diem_root_address(), None, None),
+        MethodRequest::GetAccountStateWithProof(address, None, None),
+    ];
+    let callback: RequestCallback = |ctxt, subresponses| match subresponses {
+        [MethodResponse::GetAccountStateWithProof(ref diem_root), MethodResponse::GetAccountStateWithProof(ref account)] =>
+        {
+            let diem_root_with_proof: AccountStateWithProof = diem_root.try_into()?;
+            let account_state_with_proof: AccountStateWithProof = account.try_into()?;
+
+            let latest_li = ctxt.state_proof.0.ledger_info();
+            let state_version = latest_li.version();
+
+            let address = match ctxt.request {
+                MethodRequest::GetAccount((address,)) => *address,
+                _ => panic!("should not happen"),
+            };
+
+            diem_root_with_proof.verify(latest_li, state_version, diem_root_address())?;
+            account_state_with_proof.verify(latest_li, state_version, address)?;
+
+            // TODO(philiphayes): it seems wasteful to lookup the whole diem_root
+            // account, verify the proofs, etc... just to get a list of supported
+            // currency codes. Would it make sense for the AccountView to just
+            // list _all_ BalanceResource's under balance? Then we could avoid
+            // this extra lookup. It's even more problematic when a batch
+            // includes many GetAccount requests, as this batch handling logic
+            // isn't (currently) smart enough to dedup the diem_root lookups.
+            let diem_root_blob = diem_root_with_proof
+                .blob
+                .ok_or_else(|| format_err!("missing diem_root accound somehow"))?;
+            let diem_root = AccountState::try_from(&diem_root_blob)?;
+            let currencies: Option<RegisteredCurrencies> = diem_root.get_config()?;
+            let currency_codes = currencies
+                .as_ref()
+                .map(RegisteredCurrencies::currency_codes)
+                .ok_or_else(|| format_err!("diem_root missing registered currencies"))?;
+
+            let maybe_account_view = account_state_with_proof
+                .blob
+                .map(|blob| {
+                    let account_state = AccountState::try_from(&blob)?;
+                    AccountView::try_from_account_state(
+                        address,
+                        account_state,
+                        &currency_codes,
+                        state_version,
+                    )
+                })
+                .transpose()?;
+
+            Ok(MethodResponse::GetAccount(maybe_account_view))
+        }
+        _ => bail!("invalid responses"),
+    };
+    VerifyingRequest::new(request, subrequests, callback)
+}
+
+// TODO(philiphayes): reuse secure/storage?
+
+pub trait Storage: Debug {
+    fn get(&self, key: &str) -> Result<Vec<u8>>;
+    fn set(&mut self, key: &str, value: Vec<u8>) -> Result<()>;
+}
```

Comment:
I think that maybe we should Seal this trait for the time being and only provide a small list of supported storages

Reply:
done